### PR TITLE
✨  Source Instagram: Migrate Instagram sources Media, User, UserLiftimelnsights and Stories to low-code

### PR DESCRIPTION
**Pull request by aldogonzalez8**
## What
We want to migrate certified connectors to low-code; in this case, it is Instagram Connector. We are initially migrating the following streams:

- Media
- Stories
- Users
- UserLifetimeInsights

Follow up in the next PR:
- StortInsights
- MediaInsights

**Important**: The original implementation [used an api object](https://github.com/airbytehq/airbyte/blob/46b2188fe05d149adf4be4a2378363ee07aa95ff/airbyte-integrations/connectors/source-instagram/source_instagram/api.py#L80) to slice by user business accounts on [all the Streams](https://github.com/airbytehq/airbyte/blob/46b2188fe05d149adf4be4a2378363ee07aa95ff/airbyte-integrations/connectors/source-instagram/source_instagram/streams.py#L77). To achieve this functionality I have created a new Api source that gets this information as pass it to other streams as a parent so they will slice by each account

## How
I have followed [this guide](https://docs.google.com/document/d/1pE_PXBjb1NWh_l6hvV7YXMda_Dx0xRRbWwZd6phMzGg/edit).

TLDR is the following

- Fill [this template](https://docs.google.com/spreadsheets/d/1BxdlZpKbdiGAYPUCAl8od-H0mLaEbpQ2NfBBBDickq4/edit#gid=0) with information about the connector.
- Model the streams through connector builder.
- If some feature is missing, add a custom component.
- Update base code:
   1. Put the manifest
   2. update source file
   3. add testing around custom components. 
   4. Run CAT

## Review guide
1. Manifest.yml: The manifest now includes the streams and schemas, so schemas for the scope of this PR were deleted.
2. Source.py: Source was updated to YamlDeclarativeSource, and there is an override on streams to include non-migrated streams
3. components.py: custom components created to handle:
   a. Fetch data for children field of Media; children are an array of media IDs that require their own data, so we go through the list to update the original children's object.
   b. Clear URL params "_nc_rid" and "ccb"
   c. convert breakdown list to an object
4. tests_components.py: test for the new components
5. streams.py: removed the old Python streams in the scope of this PR and updated parent classes of left streams.

## Can this PR be safely reverted and rolled back?
- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by aldogonzalez8** [comment]
Acceptance tests results

image

File: airbyte-integrations/connectors/source-instagram/acceptance-test-config.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ test_strictness_level: high
 acceptance_tests:
   spec:
     tests:
-      - spec_path: "integration_tests/spec.json"
+      - spec_path: "spec.json"
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-instagram/acceptance-test-config.yml:8]: Does this file actually exist? I've seen @btkcodedev for some reason rename jsons to yamls, and I have no clue why he did that, seems sus.

File: airbyte-integrations/connectors/source-instagram/metadata.yaml [hunk 2]
```diff
@@ -7,7 +7,7 @@ data:
   connectorSubtype: api
   connectorType: source
   definitionId: 6acf6b55-4f1e-4fca-944e-1a3caef8aba8
-  dockerImageTag: 3.0.10
+  dockerImageTag: 3.0.11
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-instagram/metadata.yaml:10]: @maxi297 @bazarnov what's your take on version bump when we're completely rewriting the connector? Should this be `4.0.0`? I get that it's compatible, buuuuut.

File: airbyte-integrations/connectors/source-instagram/pyproject.toml [hunk 3]
```diff
@@ -18,7 +18,7 @@ include = "source_instagram"
 [tool.poetry.dependencies]
 python = "^3.9,<3.12"
 facebook-business = "==19.0.3"
-airbyte-cdk = "0.80.0"
+airbyte-cdk = "1.3.2"
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-instagram/pyproject.toml:21]: This could be internal policy for high-usage connectors to pin specifically. But, what if you `^1` here?

File: airbyte-integrations/connectors/source-instagram/source_instagram/components.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+
+@dataclass
+class InstagramBreakDownResultsTransformation(RecordTransformation):
+    """
+    Converts an array of breakdowns into an object.
+    """
+
+    def transform(self, record: MutableMapping[str, Any], **kwargs) -> MutableMapping[str, Any]:
+        record["value"] = {res.get("dimension_values")[0]: res.get("value") for res in record["value"]}
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-instagram/source_instagram/components.py:72]: Oh interesting, is this generally just an `unnest` or `flatten` transformation? I wonder if we have other APIs where we need to do this? Is this a candidate for a low-code feature? /cc @maxi297 @tolik0 @artem1205

File: airbyte-integrations/connectors/source-instagram/source_instagram/manifest.yaml [hunk 5]
```diff
@@ -0,0 +1,732 @@
+version: 1.3.1
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-instagram/source_instagram/manifest.yaml:1]: @btkcodedev see? When you make a connector in Builder, it's supposed to use the new CDK ;)

Comment (btkcodedev) [reply airbyte-integrations/connectors/source-instagram/source_instagram/manifest.yaml:1]: Sure! Will prune Docker volumes and fetch new for next one

File: airbyte-integrations/connectors/source-instagram/acceptance-test-config.yml (hunk 1 above)
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-instagram/acceptance-test-config.yml]: Updated, but this still seems not to have an effect at AT at all.

File: airbyte-integrations/connectors/source-instagram/pyproject.toml (hunk 3 above)
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-instagram/pyproject.toml]: Done.

**Comment by aldogonzalez8** [comment]
For live tests, here you can find an analysis of the failing test **test_catalog_are_the_same**

image

https://www.loom.com/share/3d994c40315c4dc78cfb6f2f904da1d1
[catalog_diff_iterable_item_added.json](https://github.com/user-attachments/files/15857635/catalog_diff_iterable_item_added.json)
[catalog_diff_iterable_item_removed.json](https://github.com/user-attachments/files/15857636/catalog_diff_iterable_item_removed.json)

Cc @maxi297 in case you are reviewing and saw those results

File: airbyte-integrations/connectors/source-instagram/source_instagram/components.py (hunk 4 above)
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-instagram/source_instagram/components.py]: @natikgadzhi It flattens a nested array of breakdown results. I have added a better description to the method. Thanks for bringing this to my attention.

```
    The transformation flattens a nested array of breakdown results located at total_value.breakdowns[0].results into a single object
    (dictionary). In this transformation, each key-value pair in the resulting object represents a dimension and its corresponding value.
    e.g.
        {
        "name": "follower_demographics",
        "period": "lifetime",
        "title": "Follower demographics",
        "description": "The demographic characteristics of followers, including countries, cities and gender distribution.",
        "total_value": {
          "breakdowns": [
            {
              "dimension_keys": [
                "city"
              ],
              "results": [
                {
                  "dimension_values": [
                    "London, England"
                  ],
                  "value": 263
                },
                {
                  "dimension_values": [
                    "Sydney, New South Wales"
                  ],
                  "value": 467
                }
              ]
            }
          ]
        },
        "id": "17841457631192237/insights/follower_demographics/lifetime"
      }
    to:
        {
        "value": {
          "London, England": 263,
          "Sydney, New South Wales": 467,
        }
      }
```

File: airbyte-integrations/connectors/source-instagram/metadata.yaml (hunk 2 above)
Comment (aldogonzalez8) [reply airbyte-integrations/connectors/source-instagram/metadata.yaml:10]: I've run live tests on a few connections without a problem. I don't know if this implies that my changes wouldn't need to be breaking changes. The new thing is the Api parent Stream ([description](https://github.com/airbytehq/airbyte/pull/39504#issue-2353878097)), which passes the buz accounts to (now) child Streams. Anyway, I will hear experience call on this matter.

image

Comment (maxi297) [reply airbyte-integrations/connectors/source-instagram/metadata.yaml:10]: If the state, config and record outputs all have the same format and it does not add relatively big features, I would consider this a patch

Comment (bazarnov) [reply airbyte-integrations/connectors/source-instagram/metadata.yaml:10]: Once there are no regressions in terms of a breaking change, there is no need to make a `major` update, but the `patch`. @aldogonzalez8 FYI

**Comment by aldogonzalez8** [comment]
Today's run from live tests looks good, sharing some screenshots.

image
image
image
image
image

**Review by bazarnov — APPROVED** [review]
LGTM! Nice!

**Comment by aldogonzalez8** [comment]
**This will be merged on June 24th, early in the Morning.**
